### PR TITLE
Input redirection doesn't work with powershell

### PR DESCRIPTION
**Pull request by yegappan**
Patch contributed by Enan Ajmain.

### CODE REVIEW

File: src/ex_cmds.c [hunk 1]
```diff
@@ -1531,26 +1531,47 @@ make_filter_cmd(
...
+    {
+	is_powershell = (shell_name[0] == 'p')
+			&& (fnamecmp(shell_name, "powershell") == 0
+				|| fnamecmp(shell_name, "powershell.exe") == 0
+				|| fnamecmp(shell_name, "pwsh") == 0
+				|| fnamecmp(shell_name, "pwsh.exe") == 0);
+	if (is_powershell)
+	    len = (long_u)STRLEN(cmd) + 3;
+	else
+	    len = (long_u)STRLEN(cmd) + 3;			// "()" + NUL
```
Comment (3N4N) [inline src/ex_cmds.c]: Check for powershell looks redundant.

File: src/ex_cmds.c [hunk 2]
```diff
@@ -1531,26 +1531,47 @@ make_filter_cmd(
...
     // Account for fish's different syntax for subshells
     is_fish_shell = (fnamecmp(shell_name, "fish") == 0);
-    vim_free(shell_name);
     if (is_fish_shell)
 	len = (long_u)STRLEN(cmd) + 13;		// "begin; " + "; end" + NUL
     else
 #endif
-	len = (long_u)STRLEN(cmd) + 3;			// "()" + NUL
+    {
+	is_powershell = (shell_name[0] == 'p')
```
Comment (3N4N) [inline src/ex_cmds.c:1551]: PowerShell is nowadays available for Linux, too, in the form of PowerShell Core (v7). (Ignore if we don't care about that too much; who would use powershell in Linux anyway?)

File: src/ex_cmds.c [hunk 3]
```diff
@@ -1531,26 +1531,47 @@ make_filter_cmd(
...
+	else
+	    len = (long_u)STRLEN(cmd) + 3;			// "()" + NUL
+    }
+
     if (itmp != NULL)
-	len += (long_u)STRLEN(itmp) + 9;		// " { < " + " } "
+    {
+	if (is_powershell)
+	    // "& { Get-Content" + " | & " + " }"
+	    len += (long_u)STRLEN(itmp) + 23;
```
Comment (3N4N) [inline src/ex_cmds.c]: There is a space after `Get-Content` (see L1603). Addition to `len` should be 24, I think.

File: src/ex_cmds.c (hunk 2 above)
Comment (3N4N) [reply src/ex_cmds.c:1551]: Never mind. I misread `#endif` as `#else`.

File: src/ex_cmds.c [hunk 4]
```diff
@@ -1575,45 +1596,61 @@ make_filter_cmd(
 	STRCAT(buf, itmp);
     }
 #else
-    // For shells that don't understand braces around commands, at least allow
-    // the use of commands in a pipe.
-    if (*p_sxe != NUL && *p_sxq == '(')
+    if (is_powershell)
```
Comment (3N4N) [inline src/ex_cmds.c]: Okay, here is where I wanted to mention powershell being available for Linux, too (mistakenly did so on a previous line). Taking the `if is_powershell` block _above_ `#ifdef UNIX` should be enough. (Unless we don't care for people using PowerShell on Linux, assuming anyone does.)

File: src/ex_cmds.c (hunk 1 above)
Comment (yegappan) [reply src/ex_cmds.c]: Yes.  In the original patch, the length was set to zero.  But that broke some other functionality.  When I reverted that change, I forgot to remove this.

File: src/ex_cmds.c (hunk 3 above)
Comment (yegappan) [reply src/ex_cmds.c]: Changed the value to 24.

File: src/ex_cmds.c (hunk 4 above)
Comment (yegappan) [reply src/ex_cmds.c]: Moved the powershell handling code to outside of the '#ifdef/#endif'.

**Comment by 3N4N** [comment]
The patch now looks like it should be fine. I can't tell why the tests are failing (haven't looked at the test files yet).

**Comment by brammool** [comment]
The current patch checks for powershell on any system.
It is available on Linux, but does it use exactly the same syntax then?
Is it even practical to set 'shell' to powershell on any other system
than MS-Windows?

-- 
hundred-and-one symptoms of being an internet addict:
256. You are able to write down over 250 symptoms of being an internet
     addict, even though they only asked for 101.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by 3N4N** [comment]
> It is available on Linux, but does it use exactly the same syntax then?

There may be differences in some cases, but as far as I have used it I haven't found any.

> Is it even practical to set 'shell' to powershell on any other system
> than MS-Windows?

I don't think so. First, practically no one will use it. Second, there are several bugs/oddities in pwsh in Linux: e.g., using `sort` with `Start-Process sort -RedirectStandardInput in.txt` works fine, but if we redirect it to a file (`-RedirectStandardOutput out.txt`) then it eats all the empty lines.

But I want to leave it to the maintainer (you, Bram) to decide. We can always take the other choice later, if required; it's not a huge change.

**Comment by vim-ml** [comment]
Hi Bram,

On Mon, Oct 3, 2022 at 5:06 AM Bram Moolenaar ***@***.***>
wrote:

>
> The current patch checks for powershell on any system.
> It is available on Linux, but does it use exactly the same syntax then?
> Is it even practical to set 'shell' to powershell on any other system
> than MS-Windows?
>
>
>
Setting 'shell' to 'pwsh' or 'powershell' (Powershell) was supported on
Linux before
this patch.  But "%!sort" was failing on Linux (just like on MS-Windows)
before
this patch.  This patch makes it work on both Linux and MS-Windows.

- Yegappan

**Pull request closed without merging**